lib/libzfs/libzfs_mount.c

```diff
@@ -1302,12 +1302,13 @@ mountpoint_cmp(const void *arga, const void *argb)
 }
 
 /*
- * Return true if path2 is a child of path1.
+ * Return true if path2 is a child of path1 or path2 equals path1.
  */
 static boolean_t
 libzfs_path_contains(const char *path1, const char *path2)
 {
-	return (strstr(path2, path1) == path2 && path2[strlen(path1)] == '/');
+	return (strcmp(path1, path2) == 0) ||
```

Comment:
Can you explain how this fixes the issue at hand? If we have a list of mountpoints like this:
```
1. /a (canmount=off)
2. /a
3. /a/b
4. /a/c
```
How will this prevent the thread mounting (1) (which incidentally won't mount `/a`) from dispatching mounts for (3) and (4)?

Reply:
If `libzfs_path_contains()` return True for the same paths (i.e. two same top level dirs on initial dispatching), then there's no chance race can happen because they never go in multi-thread (read `zfs_foreach_mountpoint()`).